File: cpp_solutions/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        int a =0;
        int b =0;
        int c=0;
        for (char ch: s){
            if (ch == '('){
                a+=1;
            }
            else if(ch == ')'){
                a-=1;
            }
            else if(ch == '{'){
                b+=1;
            }
            else if(ch == '}'){
                b-=1;
            }
            else if(ch == '['){
                c+=1;
            }
            else{
                c-=1;
            }
        }
        ////////////////////////////
        for(char ch: s){
            
            if (ch == '(' || ch =='{' || ch =='['){
                stack.push(ch);
            }
            else{
                
                if (stack.empty()){
                    return false;
                    
                }
                
                if (ch == ')'){
                    
                    char top = 'a';
                    while (!stack.empty() && top!='('){
                        top = stack.top();
                        stack.pop();
                    }
                    
                    if (top!='('){
                        return false;
                    }
                }
                
                else if(ch == '}'){
                    char top = 'a';
                    while (!stack.empty() && top!='{'){
                        top = stack.top();
                        stack.pop();
                    }
                    
                    if (top!='{'){
                        return false;
                    }
                }
                
                else{
                    
                    char top = 'a';
                    while (!stack.empty() && top!='['){
                        top = stack.top();
                        stack.pop();
                    }
                    
                    if (top!='['){
                        return false;
                    }
                    
                }
            }
        }
        /////////////////
        
        return (stack.empty() && a==0 && b==0 && c==0);
    }
};

File: python3_solution_set0/20. Valid Parentheses.py
##ss
class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = []
        
        #visited = []
        if s.count(")")!=s.count("(") or s.count("[")!=s.count("]") or s.count("{")!=s.count("}"):
            return False
        for x in range(len(s)):
            #print(stack)
            if s[x] == "]" and "[" not in stack:
                return False
            
            elif s[x] == ")" and "(" not in stack:
                #print("(" in stack)
                return False
            
            elif s[x] == "}" and "{" not in stack:
                return False
            
            
            elif s[x] == "]" and "[" in stack:
                head  = -1
                while head!="[" and len(stack)!=0:
                    head = stack.pop()
                    #print(head)
                    
                if head!="[":
                    return False
                
            elif s[x] == ")" and "(" in stack:
                head  = -1
                while head!="(" and len(stack)!=0:
            
                    head = stack.pop()
                   
                if head!="(":
                    return False
                
            elif s[x] == "}" and "{" in stack:
                head  = -1
                while head!="{" and len(stack)!=0:
                    head = stack.pop()
                    
                if head!="{":
                    return False
                
            else:
                stack.append(s[x])
                
                
        if len(stack) == 0:
            return True
        else:
            return False
                    
        
